### PR TITLE
Add SDK workspace create and deletion commands

### PR DESCRIPTION
**Pull request by ismirlia**
JIRA: https://jsw.ibm.com/browse/PPC-3726

### CODE REVIEW

File: helpers/constants.go [hunk 1]
```diff
@@ -223,6 +223,12 @@ const (
 	PIPlacementGroupPolicy = "pi_placement_group_policy"
 	PIPlacementGroupID     = "pi_placement_group_id"
```
Comment (yussufsh) [inline helpers/constants.go:225]: Let's avoid adding Terraform related const here.

File: helpers/constants.go [hunk 2]
```diff
@@ -223,6 +223,12 @@ const (
...
 
+	// IBM PI Workspace
```
Comment (yussufsh) [inline helpers/constants.go:226]: Let's avoid adding Terraform related const here.

Comment (yussufsh) [inline helpers/constants.go:226]: Let's avoid adding Terraform related const here.

**Review by yussufsh — CHANGES_REQUESTED** [review]

File: helpers/constants.go (hunk 2 above)
Comment (ismirlia) [reply helpers/constants.go:226]: I cannot.

Comment (yussufsh) [reply helpers/constants.go:226]: No, we have stopped that practice. terraform variables are used there Itself. These are older vars that can also be moved there.
Doesn't make sense to relate this repo with Terraform project only.

**Review by johnnguyen3196 — APPROVED** [review]

**Review by michaelkad — APPROVED** [review]
LGTM

File: helpers/constants.go (hunk 2 above)
Comment (michaelkad) [reply helpers/constants.go]: @yussufsh  it has been moved. Could you review ?

File: clients/instance/ibm-pi-workspaces.go [hunk 3]
```diff
@@ -54,3 +70,58 @@ func (f *IBMPIWorkspacesClient) GetAll() (*models.Workspaces, error) {
...
+		return response, fmt.Errorf("error deleting workspace: response %v, err %v", response, err)
+	}
+	if response.StatusCode >= 400 {
+		return response, fmt.Errorf("error deleting resource instance. Status code: %d", response.StatusCode)
+	}
+	return response, nil
+}
+
+// Get a resourceController
+func (f *IBMPIWorkspacesClient) GetRC(rcWorkspaceID string) (*resourcecontrollerv2.ResourceInstance, *core.DetailedResponse, error) {
```
Comment (yussufsh) [inline clients/instance/ibm-pi-workspaces.go:113]:
```suggestion
// Get a workspace
func (f *IBMPIWorkspacesClient) Get(workspaceID string) (*resourcecontrollerv2.ResourceInstance, *core.DetailedResponse, error) {
```

File: clients/instance/ibm-pi-workspaces.go [hunk 4]
```diff
@@ -54,3 +70,58 @@ func (f *IBMPIWorkspacesClient) GetAll() (*models.Workspaces, error) {
...
+	resourceController, err := ibmpisession.CreateResourceControllerV2(f.session.Options.URL, f.session.Options.Authenticator)
+	if err != nil {
+		return nil, nil, fmt.Errorf("error creating Resource Controller client: %v", err)
+	}
+	planID := translatePlan(plan)
+	if planID == "" {
+		return nil, nil, fmt.Errorf("workspace creation error, incorrect plan value; either \"public\" or \"private\" is allowed")
+	}
+	params := resourceController.NewCreateResourceInstanceOptions(name, location, groupID, planID)
+	controller, response, err := resourceController.CreateResourceInstance(params)
```
Comment (yussufsh) [inline clients/instance/ibm-pi-workspaces.go:85]: Suggest re-naming `controller` -> `workspace` or `cloud_instance` everywhere.

File: ibmpisession/utils.go [hunk 5]
```diff
@@ -106,3 +107,19 @@ func SDKFailWithAPIError(err error, origErr error) error {
 	}
 	return origErr
 }
+
+// CreateResourceControllerV2  returns a resourceControllerV2
+func CreateResourceControllerV2(url string, authenticator core.Authenticator) (service *rc.ResourceControllerV2, err error) {
+	endpoint := url
```
Comment (yussufsh) [inline ibmpisession/utils.go:113]: Why using this new variable? Use `url` directly.

**Review by yussufsh — CHANGES_REQUESTED** [review]

**Comment by yussufsh** [comment]
@ismirlia Could you also add workspace in `examples` as well? If this feature is urgent then it can go in a different PR (please create an issue in that case).

File: clients/instance/ibm-pi-workspaces.go [hunk 6]
```diff
@@ -54,3 +70,58 @@ func (f *IBMPIWorkspacesClient) GetAll() (*models.Workspaces, error) {
...
+	if err != nil {
+		return nil, response, fmt.Errorf("error creating workspace: controller %v response %v  err %v", controller, response, err)
+	}
+	if response.StatusCode >= 400 {
```
Comment (yussufsh) [inline clients/instance/ibm-pi-workspaces.go:89]: Don't we also want to use `ibmpisession.SDKFailWithAPIError`?

File: clients/instance/ibm-pi-workspaces.go (hunk 3 above)
Comment (yussufsh) [reply clients/instance/ibm-pi-workspaces.go:113]: Why different `Get` approaches? Do they give different data with RC?

Comment (ismirlia) [reply clients/instance/ibm-pi-workspaces.go:113]: Yes the Get is the powervs workspace get endpoint, and the Get resource controller is the resource controller's get endpoint. They return slightly different data, and the our get endpoint is only implement in powervs not stratos.

Comment (yussufsh) [reply clients/instance/ibm-pi-workspaces.go:113]: Do we need both versions?

File: clients/instance/ibm-pi-workspaces.go (hunk 6 above)
Comment (ismirlia) [reply clients/instance/ibm-pi-workspaces.go:89]: Possibly? I'm implementing calls from a different SDK, and I didn't want to introduce new potential errors. I would prefer to revise this after the break.

**Comment by ismirlia** [comment]
@michaelkad Could you open the ticket to create an example?

**Comment by yussufsh** [comment]
> @michaelkad Could you open the ticket to create an example?

Created #273 
Please address unresolved conversations and mark them resolved. No more review comments.

**Comment by michaelkad** [comment]
> > @michaelkad Could you open the ticket to create an example?
> 
> 
> 
> Created #273 
> 
> Please address unresolved conversations and mark them resolved. No more review comments.

Yes, will do. This is urgent-- will address unresolved comments in a #273

**Pull request merged**